### PR TITLE
Fix to #10785 - Model Query Filter generates incorrect SQL

### PR DESCRIPTION
**Pull request by maumar**
Problem was that we would not apply null semantics properly to ColumnReference expressions - we were not detecting that the column behind it is actually nullable and IS NULL term needs to be generated for it.
Also as part of this change improved test infrastructure for cases when some of the results are null.

### CODE REVIEW

File: src/EFCore.Specification.Tests/Query/GearsOfWarQueryFixtureBase.cs [hunk 1]
```diff
@@ -38,120 +38,180 @@ protected GearsOfWarQueryFixtureBase()
                     typeof(City),
                     (e, a) =>
                         {
-                            Assert.Equal(e.Name, a.Name);
-                            Assert.Equal(e.Location, a.Location);
+                            Assert.Equal(e == null, a == null);
```
Comment (smitpatel) [inline src/EFCore.Specification.Tests/Query/GearsOfWarQueryFixtureBase.cs]: These all could be simplified by writing `Equals` method on Class directly.

**Comment by maumar** [comment]
merged 154725799c44ac14e7592c17c2655037fcf76261
